crates/net/network/src/transactions/mod.rs

```diff
@@ -1050,149 +1065,174 @@ where
+                                this.on_bad_import(err.hash);
+                                continue
                             }
+                            this.on_good_import(err.hash);
                         }
-                        Poll::Ready(None) => break, // todo: handle stream closed as error
-                        Poll::Pending => break,
                     }
                 }
-            };
-        }
+            }
+        );
 
-        loop {
-            // drain network/peer related events
-            poll_nested_stream_with_yield_points!(
-                1,
-                this.network_events.poll_next_unpin(cx),
-                |event| this.on_network_event(event)
-            );
+        // advance network/peer related events (update peers)
+        let maybe_more_network_events = poll_nested_stream_with_yield_points!(
+            "net::tx",
+            "Network events",
+            DEFAULT_BUDGET_TRY_DRAIN_STREAM,
+            this.network_events.poll_next_unpin(cx),
+            |event| this.on_network_event(event)
+        );
 
-            if this.has_capacity_for_fetching_pending_hashes() {
-                // try drain buffered transactions.
-                let info = &this.pending_pool_imports_info;
-                let max_pending_pool_imports = info.max_pending_pool_imports;
-                let has_capacity_wrt_pending_pool_imports =
-                    |divisor| info.has_capacity(max_pending_pool_imports / divisor);
-
-                let metrics = &this.metrics;
-                let metrics_increment_egress_peer_channel_full =
-                    || metrics.egress_peer_channel_full.increment(1);
-
-                this.transaction_fetcher.on_fetch_pending_hashes(
-                    &this.peers,
-                    has_capacity_wrt_pending_pool_imports,
-                    metrics_increment_egress_peer_channel_full,
-                );
-            }
+        // try drain successfully imported transactions to propagate (inform peers which txns
+        // we have seen)
+        let mut new_txs = Vec::new();
 
-            // drain commands
-            poll_nested_stream_with_yield_points!(1, this.command_rx.poll_next_unpin(cx), |cmd| {
-                this.on_command(cmd)
-            });
+        let maybe_more_pending_txns = poll_nested_stream_with_yield_points!(
+            "net::tx",
+            "Pending transactions stream",
+            DEFAULT_BUDGET_TRY_DRAIN_POOL_IMPORTS,
+            this.imported_transactions.poll_next_unpin(cx),
+            |hash| new_txs.push(hash)
+        );
 
-            // drain incoming transaction events
-            poll_nested_stream_with_yield_points!(
-                1,
-                this.transaction_events.poll_next_unpin(cx),
-                |event| this.on_network_tx_event(event)
-            );
+        if !new_txs.is_empty() {
+            this.on_new_transactions(new_txs);
+        }
```

Comment:
I don't think we need any preemption here

because we want to emit all new pending txs, and this number will naturally be very low

Reply:
can you guarantee this theory in a scenario with byzantine nodes? I'm benchmarking this atm